Source/JavaScriptCore/assembler/MacroAssemblerX86_64.h

```diff
@@ -3297,15 +3297,97 @@ class MacroAssemblerX86_64 : public MacroAssemblerX86Common {
-    void vectorLoad32Lane(Address address, TrustedImm32 imm, FPRegisterID dest) { UNUSED_PARAM(address); UNUSED_PARAM(imm); UNUSED_PARAM(dest); }
-    void vectorLoad64Lane(Address address, TrustedImm32 imm, FPRegisterID dest) { UNUSED_PARAM(address); UNUSED_PARAM(imm); UNUSED_PARAM(dest); }
+    void vectorLoad8Splat(Address address, FPRegisterID dest, FPRegisterID scratch)
+    {
+        if (supportsAVXForSIMD()) {
+            m_assembler.vpinsrb_mrr(0, address.offset, address.base, dest, dest);
+            m_assembler.vpxor_rrr(scratch, scratch, scratch);
+            m_assembler.vpshufb_rrr(scratch, dest, dest);
+        } else
+            RELEASE_ASSERT_NOT_REACHED();
+    }
+
+    void vectorLoad16Splat(Address address, FPRegisterID dest)
+    {
+        if (supportsAVXForSIMD()) {
+            m_assembler.vpinsrw_mrr(0, address.offset, address.base, dest, dest);
+            m_assembler.pshuflw_rr(0, dest, dest);
+            m_assembler.punpcklqdq_rr(dest, dest);
+        } else
+            RELEASE_ASSERT_NOT_REACHED();
+    }
+
+    void vectorLoad32Splat(Address address, FPRegisterID dest)
+    {
+        if (supportsAVXForSIMD())
+            m_assembler.vbroadcast_mr(address.offset, address.base, dest);
+        else
+            RELEASE_ASSERT_NOT_REACHED();
+    }
+
+    void vectorLoad64Splat(Address address, FPRegisterID dest)
+    {
+        if (supportsAVXForSIMD())
+            m_assembler.vmovddup_mr(address.offset, address.base, dest);
+        else
+            RELEASE_ASSERT_NOT_REACHED();
+    }
+
+    void vectorLoad8Lane(Address address, TrustedImm32 imm, FPRegisterID dest)
+    {
+        if (supportsAVXForSIMD())
+            m_assembler.vpinsrb_mrr(imm.m_value, address.offset, address.base, dest, dest);
+        else
+            RELEASE_ASSERT_NOT_REACHED();
+    }
+
+    void vectorLoad16Lane(Address address, TrustedImm32 imm, FPRegisterID dest)
+    {
+        if (supportsAVXForSIMD())
+            m_assembler.vpinsrw_mrr(imm.m_value, address.offset, address.base, dest, dest);
+        else
+            RELEASE_ASSERT_NOT_REACHED();
+    }
+
+    void vectorLoad32Lane(Address address, TrustedImm32 imm, FPRegisterID dest)
+    {
+        if (supportsAVXForSIMD())
+            m_assembler.vpinsrd_mrr(imm.m_value, address.offset, address.base, dest, dest);
+        else
+            RELEASE_ASSERT_NOT_REACHED();
+    }
+
+    void vectorLoad64Lane(Address address, TrustedImm32 imm, FPRegisterID dest)
+    {
+        if (supportsAVXForSIMD())
+            m_assembler.vpinsrq_mrr(imm.m_value, address.offset, address.base, dest, dest);
+        else
+            RELEASE_ASSERT_NOT_REACHED();
+    }
+
+    void vectorStore8Lane(TrustedImm32 imm, FPRegisterID src, Address address, RegisterID scratch)
+    {
+        vectorExtractLane(SIMDLane::i8x16, SIMDSignMode::Unsigned, imm, src, scratch);
+        store8(scratch, address);
```

Comment:
Do we need these two separate things? I think vpextrb etc. can take memory destination, so we can remove scratch etc. too. and efficiently implement them.

Reply:
Like, defining `vpextrb_rm` and use it.